# MACHINE LEARNING AND STATISTICS TASKS

## TASK 1

Write a Python function called sqrt2 that calculates and
prints to the screen the square root of 2 to 100 decimal places. Your code should
not depend on any module from the standard library1 or otherwise. You should
research the task first and include references and a description of your algorithm.

### RESEARCH

The square root of a number $S$ is a number when multiplied by itself yields the number $S$.

The most common methods to calculating the square root of a number are iterative, the principle involves two steps below:

* Find a suitable starting seed value or guess.
* Refine this value using an iterative calculation until a criteria is met to terminate.

#### Initial Seed Value Estimation

The initial seed value must be a positive non-zero number, it should be between 1 and $S$ the number we are trying to find the square root off. There are many methods of estimating this number because the better the estimate the less iterations the algorithm will have to perform to find the square root.

For estimating the seed value $x$ for approximating the square root of $ S = 2$ and the fact that the number $x$ should be 0 < $x$ < 2 then we will choose 1. 


The most common of this type of methods is the Newton Rampson approximation. This uses a mathematical formula to approximate the square root of a number this formula is run many times or iterations to get the square root approximation. The formula is given below.



Reference: https://medium.com/@surajregmi/how-to-calculate-the-square-root-of-a-number-newton-raphson-method-f8007714f64

To get the answer required the Python math function is used below to get the answer we need to approximate.




From the question above the task involved writing a program to get the square root of the number two without using any of the Python library functions such as math.sqrt. The math.sqrt function can be used to check the answer to the final solution. 

Reference: https://realpython.com/python-square-root-function/

To calculate the sqrt root of a number to a certain number of decimal places the function getcontext() can be used to set the value from defualt 28 decimal to the desired number of decimal places in this case 100

https://www.geeksforgeeks.org/setting-precision-in-python-using-decimal-module/amp/?ref=rp



In [1]:
import math as m # Import Python Math library as m
 
m.sqrt(2) #Run the function sqrt from the math library passing in 2

The Python sqrt() function without changing the number of decimal places or precision of the output only gives the answer to 16 demimal places. To change this to 100 decimal places the following code can be run

In [2]:
from decimal import * # Import decimal function
getcontext().prec = 101 #set overall to 101 to get 100 decimal places after the decimal point
ans = Decimal(2).sqrt() ## Run the Decimal function on sqrt(2) and assigned it to ans
print("The answer we need to approximate is:\n",ans) # Print the Decimal answer


There is an issue with floating point numbers if you assign the test variable to the answer we obtained with the Python maths function and print that variable we only get part of the variable i.e to 16 decimal places.

In [3]:
test = 1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727

In [4]:
print(test)

1.4142135623730951


To get a square root of a number to 100 decimal places you need to multiple the number by 10 ** 200 this will yield an integer variable for usingin the calcaluation

https://stackoverflow.com/questions/64295245/how-to-get-the-square-root-of-a-number-to-100-decimal-places-without-using-any-l

Newton Rampson approximation formula is given by the following formula

guess = (guess + $S$/guess)/2

This calculation is run repeatly until the difference between the original number $S$ and the guess squared or guess^2 is less that 0.

The python code for this formula for this is given below


In [5]:
def sqrt2():
    
    x = 2 #Number that we want to find the square root off
    
    s = x * 10 ** 200 #This is a large enough integer so when the square root is calculated its is 100 decimal places
    
    guess = s//2 #This is the initial guess
    
    while s - (guess**2) < 0: #while the guess or approximation squared less the original number is greater than zero continue with the approximation calculation
        guess = (guess + s//guess)//2
    
    guess = str(guess) # Convert varible to string
    print(guess[0]+"."+guess[1:]) # Print out string with dot in the correct place
    
  
        

In [6]:
sqrt2()

1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727


## TASK 2

The Chi-squared test for independence is a statistical
hypothesis test like a t-test. It is used to analyse whether two categorical variables
are independent. The Wikipedia article gives the table below as an example [4],
stating the Chi-squared value based on it is approximately 24.6. Use scipy.stats
to verify this value and calculate the associated p value. You should include a short
note with references justifying your analysis in a markdown cell.

<img src="Images/collars_array_table.JPG" alt="Table" style="width: 500px;"/>





### RESEARCH

To verify the chi2 test from the data given in the question the following reference was used to do a statistical analysis on the number you get when throwing a dice
Ref: https://towardsdatascience.com/running-chi-square-tests-in-python-with-die-roll-data-b9903817c51b


In [2]:
import numpy as np

# The first step is to take the data and make three arrays of four of the collar colour categories from the samples A to D

white_collar = [90, 60, 104, 95]
blue_collar = [30, 50, 51, 20]
no_collar = [30, 40, 45, 35]

collars = np.array([white_collar, blue_collar, no_collar]) #Construct the numpy array with the three arrays 

print(collars)


[[ 90  60 104  95]
 [ 30  50  51  20]
 [ 30  40  45  35]]


In [3]:
from scipy import stats

stats.chi2_contingency(collars)

(24.5712028585826,
 0.0004098425861096696,
 6,
 array([[ 80.53846154,  80.53846154, 107.38461538,  80.53846154],
        [ 34.84615385,  34.84615385,  46.46153846,  34.84615385],
        [ 34.61538462,  34.61538462,  46.15384615,  34.61538462]]))

From the output above we can see that the following values are calculated $\chi^2$, $\rho-value$ and Degrees of Freedom ($DF$)

$\chi^2$ = 24.57 or 24.6<br/>
$\rho-value$ = 0.00041<br/>
$DF$ = 6

The expected array values that is calculated is given below, if the variables were independent. This array is given below rounded to whole numbers for clarity

 array([[ 81,  81, 107,  81],
        [ 35,  35,  46,  35],
        [ 35,  35,  46,  35]]))
        
This is worked back in to the original arrays for a visual comparsion

**Original Given Values** <br/>
white_collar = [90, 60, 104, 95]<br/>
blue_collar = [30, 50, 51, 20]<br/>
no_collar = [30, 40, 45, 35]<br/>

**Expected Calculated Values if the variables were independent**<br/>
white_collar = [81, 81, 107, 81]<br/>
blue_collar = [35, 35, 46, 35]<br/>
no_collar = [35, 35, 46, 35]<br/>


The test for independence is the probably that the $\chi^2$ value 24.6 or higher is the $\rho-value$ which is 0.00041 or 0.041% which is extremely low. Therefore we would have to reject the hypothesis that the values fo the variables are independent

        

In [4]:
rh_longer = [11, 3, 8]
lh_longer = [2, 9, 14]
both_same = [12, 13, 28]

hand_foot = np.array([rh_longer, lh_longer, both_same])

stats.chi2_contingency(hand_foot)

(11.942092624356777, 0.0177871146098607, 4, array([[ 5.5 ,  5.5 , 11.  ],
        [ 6.25,  6.25, 12.5 ],
        [13.25, 13.25, 26.5 ]]))

There are two Hypoteses in the test for independence of the variables 

Independence = There is no dependency between the categorical variables in the dataset given i.e the values of the variables are independent of each other.

Dependence = There is a dependency between the values of the variables.

The Chi-square tests takes the values of the variables and calculates the expected values if the values were truly independent and then calculates the $\chi^2$, $\rho-value$ and the Degrees of Freedom ($DF$)